File: backend/src/main/java/com/vibeshare/Service/PostService.java
package com.vibeshare.Service;

import com.vibeshare.DTO.Request.CreatePostRequest;
import com.vibeshare.DTO.Response.PostResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface PostService {

    /* ================= CREATE ================= */

    PostResponse createPost(CreatePostRequest request);

    PostResponse createPostWithImage(
            Long userId,
            String content,
            String type,
            MultipartFile file
    );

    PostResponse createPostWithVideo(
            Long userId,
            String content,
            String type,
            MultipartFile file
    );

    /* ================= READ ================= */

    // Feed (viewer context)
    List<PostResponse> getFeed(Long currentUserId);

    // Profile posts (viewer + profile owner)
    List<PostResponse> getPostsByUserId(
            Long profileUserId,
            Long currentUserId
    );

    // Single post
    PostResponse getPostById(
            Long postId,
            Long currentUserId
    );

    /* ================= DELETE ================= */

    void deletePost(Long postId, Long userId);

    /* ================= LIKE / UNLIKE ================= */

    // ✅ BOTH return updated post
    PostResponse likePost(Long postId, Long userId);

    PostResponse unlikePost(Long postId, Long userId);

    PostResponse incrementViews(Long postId);

    List<PostResponse> getReels(Long currentUserId);

    List<PostResponse> getStories();
}

File: frontend/src/api/post.service.ts
import type { PostResponseType } from "@/types/PostResponseType";
import { apiFetch } from "./client";

export const postService = {
  /* ================= CREATE ================= */

  createImagePost: (content: string, file: File, type: string = "POST") => {
    const formData = new FormData();
    formData.append("content", content);
    formData.append("type", type);
    formData.append("file", file);

    return apiFetch<PostResponseType>("/api/posts/with-image", {
      method: "POST",
      body: formData,
    });
  },

  createVideoPost: (content: string, file: File, type: string = "REEL") => {
    const formData = new FormData();
    formData.append("content", content);
    formData.append("type", type);
    formData.append("file", file);

    return apiFetch<PostResponseType>("/api/posts/with-video", {
      method: "POST",
      body: formData,
    });
  },

  /* ================= READ ================= */

  getFeed: () => {
    return apiFetch<PostResponseType[]>("/api/posts/feed");
  },

  getReels: () => {
    return apiFetch<PostResponseType[]>("/api/posts/reels");
  },

  getStories: () => {
    return apiFetch<PostResponseType[]>("/api/posts/stories");
  },

  getPostByUser: (userId: number) => {
    return apiFetch<PostResponseType[]>(`/api/posts/user/${userId}`);
  },

  getPostById: (postId: number) => {
    return apiFetch<PostResponseType>(`/api/posts/${postId}`);
  },

  /* ================= LIKE / UNLIKE ================= */

  likePost: (postId: number) => {
    return apiFetch<PostResponseType>(`/api/posts/${postId}/like`, {
      method: "POST",
    });
  },

  unlikePost: (postId: number) => {
    return apiFetch<PostResponseType>(`/api/posts/${postId}/like`, {
      method: "DELETE",
    });
  },

  /* ================= VIEWS ================= */

  incrementViews: (postId: number) => {
    return apiFetch<PostResponseType>(`/api/posts/${postId}/view`, {
      method: "PUT",
    });
  },

  /* ================= DELETE ================= */

  deletePost: (postId: number) => {
    return apiFetch<void>(`/api/posts/${postId}`, {
      method: "DELETE",
    });
  },
};
